teps/0023-implicit-mapping.md

```diff
@@ -0,0 +1,358 @@
+top-most parameter definition -
+
+1. We need a param named `MESSAGE`
+2. The param value is of type `string`
+3. No default value is required since we already have a value.
+
+Since these are all embedded specs within the same resource, our expectation is
+that this shortened syntax will not take away from the context knowledge of
+parameters and how they should be used. A good analogy here is variable scoping
+found in most programming languages - e.g.
+
+```go
+x := "Hi!"
+func() {
+  // x is still reachable, even though it's not explicitly defined parameter.
+  fmt.Println(x)
+}()
+```
+
+### Admission Controller
+
+This logic should live within the Pipelines mutating admission controller.
+
+Although the resource author is using a shortcut, we should still store and
+validate a fully resolved spec. This spec should look very similar to the
+explicit config, using the information we derived:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-taskspec-to-echo-message
+spec:
+  pipelineSpec:
+    params:
+      - name: MESSAGE
+        type: string
+    tasks:
+      - name: echo-message
+        taskSpec:
+          params:
+            - name: MESSAGE
+              type: string
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+        params:
+          - name: MESSAGE
+            value: $(params.MESSAGE)
+  params:
+    - name: MESSAGE
+      value: "Good Morning!"
+```
+
+[Extra params are safe to include in Tasks](https://github.com/tektoncd/pipeline/blob/0593c7bef395505fae8b2e611cf4632e2e980e1d/examples/v1beta1/pipelineruns/pipelinerun-with-extra-params.yaml)
+since they don't modify behavior unless they are actually used - because of this
+there is little risk in passing through all parent param values to embedded
+specs.
+
+To avoid issues of additional webhook latency or reliability, we will not
+support remote TaskRefs or PipelineRefs.
+
+### Caveats
+
+#### Naming conflicts
+
+In cases of parameter naming conflicts, the innermost definition should win.
+e.g. for the following config:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-taskspec-to-echo-message
+spec:
+  pipelineSpec:
+    tasks:
+      - name: echo-message
+        taskSpec:
+          params:
+            - name: MESSAGE
+              type: string
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+  params:
+    - name: MESSAGE
+      value: ["Good Morning!"]
+```
+
+The innermost Task definition should take precedence and make the config
+**invalid**, since the PipelineRun array param should not be able to override
+the param definition of the task.
```

Comment:
what would be the expanded result here? im guessing the param would be added to the pipeline spec and then just not used?

Reply:
It would look like:

```yaml
apiVersion: tekton.dev/v1beta1
kind: PipelineRun
metadata:
  name: pipelinerun-with-taskspec-to-echo-message
spec:
  pipelineSpec:
    params:
      - name: MESSAGE
        type: array
    tasks:
      - name: echo-message
        taskSpec:
          params:
            - name: MESSAGE
              type: array
          steps:
            - name: echo
              image: ubuntu
              script: |
                #!/usr/bin/env bash
                echo "$(params.MESSAGE)"
        params:
          - name: MESSAGE
            value: $(params.MESSAGE)
  params:
    - name: MESSAGE
      value: ["Good Morning!"]
```

Which would fail validation, since we would be trying to use an array like a string.

Reply:
The array variable type would propagate all the way down to the embedded `taskSpec`, so we wouldn't hit the failure until we tried to validate the script params.

Reply:
hmm this might not be that important but @vdemeester 's example makes a bit more sense to me? in that the example taskspec above already has:

```yaml
        taskSpec:
          params:
            - name: MESSAGE
              type: string
```

(might just be a typo in your comment tho?)

regardless i think it might be useful in the TEP if you showed the expanded version as well, so it's clear why it's invalid

Reply:
d'oh! My mistake - I forgot that was included in the original example (I thought it was omitted). Added the resolved spec to the example!